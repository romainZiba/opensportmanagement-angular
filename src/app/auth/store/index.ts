export * from './login.actions';
export * from './login.state';
