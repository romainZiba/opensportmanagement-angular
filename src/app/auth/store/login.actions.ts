import { Credentials, User } from "../models/user";

export enum LoginActionsType {
  LOGIN = "[Auth] Login",
  LOGIN_SUCCESS = "[Auth] Login Success",
  LOGIN_FAILURE = "[Auth] Login Failure",
  LOGIN_REDIRECT = "[Auth] Login Redirect",
  LOGOUT = "[Auth] Logout",
  USER_LOGGED = "[Auth] User logged"
}

export class Login {
  static readonly type = LoginActionsType.LOGIN;
  constructor(public payload: Credentials) {}
}

export class LoginSuccess {
  static readonly type = LoginActionsType.LOGIN_SUCCESS;
  constructor(public payload: User) {}
}

export class LoginFailure {
  static readonly type = LoginActionsType.LOGIN_FAILURE;
  constructor(public payload: any) {}
}

export class LoginRedirect {
  static readonly type = LoginActionsType.LOGIN_REDIRECT;
}

export class Logout {
  static readonly type = LoginActionsType.LOGOUT;
}

export class UserLogged {
  static readonly type = LoginActionsType.USER_LOGGED;
  constructor(public payload: User) {}
}
