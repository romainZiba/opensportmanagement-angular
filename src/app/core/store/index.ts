import { LayoutState } from './layout.state';
import { TeamsState } from './teams.state';

export const CoreState = [LayoutState, TeamsState];

export * from './layout.actions';
export * from './layout.state';
export * from './teams.actions';
export * from './teams.state';
