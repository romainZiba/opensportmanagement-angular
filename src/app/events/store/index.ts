export * from "./events.actions";
export * from "./events.state";
