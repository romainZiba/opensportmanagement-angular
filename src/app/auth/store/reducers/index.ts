import { ActionReducerMap, createFeatureSelector } from "@ngrx/store";

import * as fromAuth from "./login.reducer";

export interface AuthState {
  status: fromAuth.State;
}

export const reducers: ActionReducerMap<AuthState> = {
  status: fromAuth.reducer
};

export const getAuthState = createFeatureSelector<AuthState>("auth");
