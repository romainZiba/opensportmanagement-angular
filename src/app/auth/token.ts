import { InjectionToken } from '@angular/core';
import { Api } from '../core/models/api';

export const AUTH_API = new InjectionToken<Api>('AUTH.api');
