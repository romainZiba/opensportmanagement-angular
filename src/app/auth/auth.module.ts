import {
  MatRadioModule,
  MatFormFieldModule,
  MatIconModule,
  MatInputModule
} from '@angular/material';
import { AUTH_API } from './token';
import { AuthHttpService } from './services/auth-http.service';
import { AuthMockService } from './services/auth-mock.service';
import { NgModule, ModuleWithProviders } from '@angular/core';
import { CommonModule } from '@angular/common';
import { LoginComponent } from './components/login.component';
import { LoginPageComponent } from './pages/login-page.component';
import { FormsModule } from '@angular/forms';
import { AuthRoutingModule } from './auth-routing.module';
import { NgxsModule } from '@ngxs/store';
import { AuthState } from './store';
import { Api } from '../core/models/api';
import { AuthService } from './services/auth.service';

@NgModule({
  imports: [
    CommonModule,
    MatRadioModule,
    MatFormFieldModule,
    MatInputModule,
    MatIconModule,
    FormsModule,
    AuthRoutingModule,
    NgxsModule.forFeature([AuthState])
  ],
  exports: [LoginComponent, LoginPageComponent],
  declarations: [LoginPageComponent, LoginComponent]
})
export class AuthModule {
  public static forRoot(authApi: Api): ModuleWithProviders {
    return {
      ngModule: AuthModule,
      providers: [
        {
          provide: AuthService,
          useClass: authApi.useMock ? AuthMockService : AuthHttpService
        },
        {
          provide: AUTH_API,
          useValue: authApi.baseUrl
        }
      ]
    };
  }
}
