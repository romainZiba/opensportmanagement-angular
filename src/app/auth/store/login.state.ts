import { User } from "../models/user";
import { Action, Selector, State, StateContext } from "@ngxs/store";
import { Router } from "@angular/router";
import { AuthService } from "../services/auth.service";
import {
  Login,
  LoginFailure,
  LoginRedirect,
  LoginSuccess,
  Logout,
  UserLogged
} from "./login.actions";
import { catchError, map, switchMap } from "rxjs/operators";
import { asapScheduler } from "rxjs/index";
import { Navigate } from "@ngxs/router-plugin";

export class AuthStateModel {
  user: User | null;
  pending: boolean;
  error: string | null;
}

@State<AuthStateModel>({
  name: "authState",
  defaults: {
    user: null,
    pending: false,
    error: null
  }
})
export class AuthState {
  constructor(private service: AuthService, private router: Router) {}

  @Selector()
  static getUser(state: AuthStateModel) {
    return state.user;
  }

  @Selector()
  static isLogged(state: AuthStateModel) {
    return !!state.user;
  }

  @Selector()
  static isLoggingIn(state: AuthStateModel) {
    return state.pending;
  }

  @Selector()
  static getError(state: AuthStateModel) {
    return state.error;
  }

  @Action(Login)
  login(
    { patchState, dispatch }: StateContext<AuthStateModel>,
    { payload }: Login
  ) {
    patchState({
      pending: true,
      error: null
    });
    return this.service.authenticate(payload.username, payload.password).pipe(
      switchMap(() => {
        return this.service
          .whoAmI()
          .pipe(
            map(
              user =>
                asapScheduler.schedule(() => dispatch(new LoginSuccess(user))),
              catchError(error => dispatch(new LoginFailure(error)))
            )
          );
      }),
      catchError(error => dispatch(new LoginFailure(error)))
    );
  }

  @Action(LoginSuccess)
  loginSuccess(
    { patchState, dispatch }: StateContext<AuthStateModel>,
    { payload }: LoginSuccess
  ) {
    patchState({
      pending: false,
      user: payload,
      error: null
    });
    dispatch(new Navigate(["/"]));
  }

  @Action(UserLogged)
  userLogged(
    { patchState, dispatch }: StateContext<AuthStateModel>,
    { payload }: UserLogged
  ) {
    patchState({
      pending: false,
      user: payload,
      error: null
    });
  }

  @Action(LoginFailure)
  loginFailed(
    { patchState }: StateContext<AuthStateModel>,
    { payload }: LoginFailure
  ) {
    patchState({
      pending: false,
      user: null,
      error: payload
    });
  }

  @Action([LoginRedirect, Logout])
  loginRedirect({ dispatch }: StateContext<AuthStateModel>) {
    dispatch(new Navigate(["/login"]));
  }
}
