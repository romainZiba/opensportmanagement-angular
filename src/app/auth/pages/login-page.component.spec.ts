import { AuthState } from './../store/login.state';
import { CookieService } from 'ngx-cookie-service';
import { TestBed, async, ComponentFixture } from '@angular/core/testing';
import { LoginPageComponent } from './login-page.component';
import { Store, NgxsModule } from '@ngxs/store';
import { AuthService } from '../services/auth.service';
import { of } from 'rxjs';

describe('LoginPageComponent', () => {
  let fixture: ComponentFixture<LoginPageComponent>;
  let component: LoginPageComponent;
  beforeEach(() => {
    TestBed.configureTestingModule({
      declarations: [LoginPageComponent],
      imports: [NgxsModule.forRoot([AuthState])],
      providers: [
        { provide: AuthService, useValue: null },
        { provide: CookieService, useValue: null }
      ]
    });
    TestBed.overrideComponent(LoginPageComponent, {
      set: {
        template: `<p>{{ loggingIn$ | async }}</p>
                <p>{{ error$ | async }}</p>'`
      }
    });
    fixture = TestBed.createComponent(LoginPageComponent);
    component = fixture.componentInstance;
    Object.defineProperty(component, 'loggingIn$', { writable: true });
    component.loggingIn$ = of(true);
    Object.defineProperty(component, 'error$', { writable: true });
    component.error$ = of('error');
    fixture.detectChanges();
  });

  it('should ne logging in', async(() => {
    expect(fixture).toMatchSnapshot();
  }));
});
