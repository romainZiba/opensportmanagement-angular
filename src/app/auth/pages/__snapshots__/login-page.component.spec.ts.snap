// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPageComponent should ne logging in 1`] = `
<app-login
  store={[Function Store]}
>
  <p>
    true
  </p><p>
    error
  </p>'
</app-login>
`;
